### DISCUSSION TITLE
skupper link create with no kube config

### DISCUSSION THREAD

**Discussion by jhr1972** [skupperproject/skupper#1142] [Q&A]
Trying to establish a link from a standalone VM into a K8S running skupper. So, on purpose I used a VM with only skupper installed and no kubeconfig available, however this does not seem to work.

My thinking is: All that skupper needs, is the token file generated from within K8S. IMHO this file contains evething needed to connect from the outside to the skupper-router: the Ingress address, certificates, etc. Thus, the skupper gateway on the VM can be setup totally agnostic on the k8s internals. However, trying to setup the gateway results in the following error:

```
$ skupper gateway init --type docker
invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable. Please point to an existing, valid kubeconfig file.
```

The same error is generated when trying to create the link directly

```
$ skupper link create secret.yaml
invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable. Please point to an existing, valid kubeconfig file.
```

Can you eliminate my missconception pls ?

**Comment by grs** [comment]
At present there are two distinct concepts relevant here: 'sites'  and 'gateways'.

A gateway is a special case deployment of the skupper router that has to be attached to a remote kubernetes based site.

A site is a deployment of skupper that can standalone and be connected or connect to other sites. There are at present two types of sites: those deployed in a kubernetes namespace and those deployed locally using podman. A podman site is an alternative way of achieving what gateways were originally intended to achieve. However at present gateways have some options not yet offered by podman sites (the reverse is also true of course).

If you are using docker (not poadman) then for now you have to stick with gateways and this means that when you create them you need to provide the context for a kubernetes namespace which is running a skupper site you want to link that gateway to.

**Comment by jhr1972** [comment]
Thx for reply. So you are saying that a podman site may feature the abilty to connect purely by token whereas a docker site does not ?

**Comment by grs** [comment]
I am saying at present there is no support for a docker site only a docker gateway. Gateways are a special case deployment that is tied to a remote site. Though gateways use tokens just like sites, the cli commands that create them hide that detail from you.

**Comment by grs** [comment]
I believe you should be able to get everything you need from a gateway however. You can use the generate-bundle subcommand to give you an artefact you can then deploy without needing the active kubecontext. See 'Creating a gateway and applying it on a different machine' in https://skupper.io/docs/cli/index.html#exposing-services-local

**Comment by omersiar** [comment]
Found this thread while having the same confusion.

In my case, VM instance would be the TCP listener and I assumed I will be fine without even a container environment (podman or docker)

Isn't it possible to have a gateway that waits connection from a OpenShift Skupper Site on bare metal? (with or without podman?)

**Comment by omersiar** [comment]
I see that there is possibility 

https://github.com/skupperproject/skupper/blob/1d113b64c9ec6ed138a10d49a430e969aa945936/client/gateway.go#L170

But the only option to get skrouterd by compiling it?
